c/parallel/src/segmented_reduce.cu

```diff
@@ -0,0 +1,470 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#include <cub/detail/choose_offset.cuh> // cub::detail::choose_offset_t
+#include <cub/detail/launcher/cuda_driver.cuh> // cub::detail::CudaDriverLauncherFactory
+#include <cub/device/dispatch/dispatch_reduce.cuh> // cub::DispatchSegmentedReduce
+#include <cub/thread/thread_load.cuh> // cub::LoadModifier
+
+#include <exception> // std::exception
+#include <format> // std::format
+#include <string> // std::string
+#include <string_view> // std::string_view
+#include <type_traits> // std::is_same_v
+#include <utility> // std::move
+
+#include "kernels/iterators.h"
+#include "kernels/operators.h"
+#include "util/context.h"
+#include "util/errors.h"
+#include "util/indirect_arg.h"
+#include "util/types.h"
+#include <cccl/c/segmented_reduce.h>
+#include <cccl/c/types.h> // cccl_type_info
+#include <nvrtc/command_list.h>
+#include <nvrtc/ltoir_list_appender.h>
+#include <stdio.h> // printf
+
+struct op_wrapper;
+struct device_segmented_reduce_policy;
+using OffsetT = unsigned long long;
+static_assert(std::is_same_v<cub::detail::choose_offset_t<OffsetT>, OffsetT>, "OffsetT must be size_t");
+
+// check we can map OffsetT to ::cuda::std::uint64_t
+static_assert(std::is_unsigned_v<OffsetT>);
+static_assert(sizeof(OffsetT) == sizeof(::cuda::std::uint64_t));
+
+struct input_iterator_t;
+struct output_iterator_t;
+struct start_offset_iterator_t;
+struct end_offset_iterator_t;
+
+namespace segmented_reduce
+{
+
+struct segmented_reduce_runtime_tuning_policy
+{
+  int block_size;
+  int items_per_thread;
+  int vector_load_length;
+
+  segmented_reduce_runtime_tuning_policy SegmentedReduce() const
+  {
+    return *this;
+  }
+
+  int ItemsPerThread() const
+  {
+    return items_per_thread;
+  }
+
+  int BlockThreads() const
+  {
+    return block_size;
+  }
+};
+
+template <typename Tuning, int N>
+Tuning find_tuning(int cc, const Tuning (&tunings)[N])
+{
+  for (const Tuning& tuning : tunings)
+  {
+    if (cc >= tuning.cc)
+    {
+      return tuning;
+    }
+  }
+
+  return tunings[N - 1];
+}
+
+struct segmented_reduce_tuning_t
+{
+  int cc;
+  int block_size;
+  int items_per_thread;
+  int vector_load_length;
+};
+
+segmented_reduce_runtime_tuning_policy get_policy(int cc, cccl_type_info accumulator_type)
+{
+  // TODO: we should update this once we figure out a way to reuse
+  // tuning logic from C++. Alternately, we should implement
+  // something better than a hardcoded default:
+  constexpr segmented_reduce_tuning_t chain[] = {{16, 256, 16, 4}, {35, 256, 20, 4}};
+
+  auto [_, block_size, items_per_thread, vector_load_length] = find_tuning(cc, chain);
+
+  // Implement part of MemBoundScaling
+  items_per_thread = CUB_MAX(1, CUB_MIN(items_per_thread * 4 / accumulator_type.size, items_per_thread * 2));
```

Comment:
Please avoid `CUB_MAX` and `CUB_MIN` since we are in the process of removing them. Just use `cuda::std::min` etc.

Reply:
Replaced `CUB_MAX` with `::cuda::std::max` and `CUB_MIN` with `::cuda_std::min`. Fortuitously, all types involved are `int`, so there is no need to specify template parameter.